common/src/types/chain_token.rs

```diff
@@ -0,0 +1,64 @@
+//! Chain-specific Token methods
+use num_bigint::BigUint;
+use util::hex::biguint_to_hex_addr;
+
+use super::{
+    chain::Chain,
+    exchange::Exchange,
+    token::{read_token_remaps, Token, USDT_TICKER, USD_TICKER},
+};
+
+impl Token {
+    /// Given an ERC-20 contract address, returns a new Token on a specific
+    /// chain.
+    pub fn from_addr_on_chain(addr: &str, chain: Chain) -> Self {
+        Self { addr: String::from(addr).to_lowercase(), chain }
+    }
+
+    /// Given an ERC-20 contract address represented as a `BigUint`, returns a
+    /// Token
+    pub fn from_addr_biguint_on_chain(addr: &BigUint, chain: Chain) -> Self {
+        Self { addr: biguint_to_hex_addr(addr).to_lowercase(), chain }
+    }
+
+    /// Given an ERC-20 contract address, returns a new Token
+    ///
+    /// Returns the first matching address found across all chains. Panics if
+    /// no matching address is found.
+    pub fn from_addr_without_chain(addr: &str) -> Self {
```

Comment:
Sometimes (like in the price reporter) given an address one must find the chain it belongs to across all configured token remaps.